### PR TITLE
Feature/metering

### PR DESCRIPTION
**Pull request by jakubneubauer**
Basic API for components metering.

### CODE REVIEW

File: api/src/main/java/org/datacleaner/api/ExecutionMeteringMessage.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    }
+
+    public ExecutionMeteringMessage putAdditionalInfo(String key, String value) {
+        if(additionalInfo == null) {additionalInfo = new HashMap<>();}
+        additionalInfo.put(key, value);
+        return this;
+    }
+
+    public Iterable<Map.Entry<String, String>> getAdditionalInfo() {
+        return additionalInfo == null ? EMPTY_ITERABLE : additionalInfo.entrySet();
```
Comment (kaspersorensen) [inline api/src/main/java/org/datacleaner/api/ExecutionMeteringMessage.java]: Instead of this null checking logic, why not just ensure that the map is always created in the constructor. Then you can also make it final and less prone to concurrency issues etc. And avoid the nasty EMPTY_ITERABLE which is just making the code more involved to read.

File: monitor/services/src/main/java/org/datacleaner/monitor/server/components/ComponentAware.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+ */
+package org.datacleaner.monitor.server.components;
+
+import org.datacleaner.api.Component;
+import org.datacleaner.descriptors.ComponentDescriptor;
+
+/**
+ * Added by jakub on 23.10.15
+ */
+public interface ComponentAware {
```
Comment (kaspersorensen) [inline monitor/services/src/main/java/org/datacleaner/monitor/server/components/ComponentAware.java]: What is this for? No javadoc and I cannot see the relevance for metering?

File: api/src/main/java/org/datacleaner/api/ExecutionMeteringMessage.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+        return type;
+    }
+
+    public ExecutionMeteringMessage putAdditionalInfo(String key, String value) {
+        if(additionalInfo == null) {additionalInfo = new HashMap<>();}
+        additionalInfo.put(key, value);
+        return this;
+    }
+
+    public Iterable<Map.Entry<String, String>> getAdditionalInfo() {
```
Comment (kaspersorensen) [inline api/src/main/java/org/datacleaner/api/ExecutionMeteringMessage.java]: Why this difficult return type instead of just returning a Map?

Then you can also use Collections.emptyMap() instead of EMPTY_ITERABLE if you disagree on my remark below :)

File: api/src/main/java/org/datacleaner/api/ExecutionMeteringMessage.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+    public ExecutionMeteringMessage(String type) {
+        this.type = type;
+    }
+
+    public String getType() {
+        return type;
+    }
+
+    public ExecutionMeteringMessage putAdditionalInfo(String key, String value) {
```
Comment (kaspersorensen) [inline api/src/main/java/org/datacleaner/api/ExecutionMeteringMessage.java]: I would discourage any mutation for a message type since these will be passed around in a highly concurrent consumer system. Rather I would expect any additional info to be provided in a constructor and made final from the start.

**Comment by kaspersorensen** [comment]
One final remark (sorry I know this is a lot for such a small PR :-)): I would love to have the message class also on master branch so that it can already now be incorporated into components.

But that also means I would like to have it in a commit without anything else - so that it can be cherry picked for multiple branches. The ComponentAware stuff is problematic in that regard.

File: api/src/main/java/org/datacleaner/api/ExecutionMeteringMessage.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+ * Boston, MA  02110-1301  USA
+ */
+package org.datacleaner.api;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Added by jakub on 22.10.15
```
Comment (kaspersorensen) [inline api/src/main/java/org/datacleaner/api/ExecutionMeteringMessage.java]: Anything in the API _must_ have proper javadoc IMO

File: api/src/main/java/org/datacleaner/api/ExecutionMeteringMessage.java [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+ */
+public class ExecutionMeteringMessage implements ComponentMessage {
```
Comment (kaspersorensen) [inline api/src/main/java/org/datacleaner/api/ExecutionMeteringMessage.java]: "Execution" seems a bit irrelevant in this regard because everything is execution. I would rather use words that really pertain to this, or none at all. For instance: UsageMeteringMessage, ServiceMeteringMessage, ComponentUsageMessage?

**Comment by jakubneubauer** [comment]
Will try it again :-)
